### ISSUE TITLE
[Issue]: CLR asserts when a code object is loaded for which no implimentation is avaialble for the GPUS in the system.

### ISSUE THREAD

**Issue by IMbackK** [ROCm-Developer-Tools/clr#102]
### Problem Description

Loading a shared object containing gpu code that is not compiled for one of the gpus in the system causes CLR to assert here https://github.com/ROCm/clr/blob/65d174c3e35423bf25c9c369513780c9d9ca760c/hipamd/src/hip_code_object.cpp#L1152 when assertions are enabled.

For release builds the assertions in clr are disabled, in this case:

When assertions are disabled and none of the gpus in the system have a implementation available in the loaded code object, clr will silently fail and the application will crash if the code object is used.

When assertions are disabled and there are gpus for which an implementation is available and gpus for which no implementation is available are present in the system clr, depending on the order in which amdgpu.ko initialized the gpus, will return without having loaded any gpu code for the code objects in question, even for the gpus that do have an implementations available.

I would like to note that i DONT think this is a bug in CLR but rather that linking against a shared library for which no gpu code is available for one of the gpus in the system is a bug in the client. However this practice has become extremely common in ROCM's libraries, mainly centered around hipBlasLT and [client code owners](https://github.com/pytorch/pytorch/issues/119081#issuecomment-2412403813) have instructed me to file a bug against clr.

hipBlasLT only supports, and thus is only compiled for, gfx908 (documentation is incorrect here), gfx90a, gfx94x and gfx11x.
In rocm, projects such as [pytorch](https://github.com/pytorch/pytorch/issues/119081) and [miopen](https://github.com/ROCm/MIOpen/pull/3030) it has become common practice to unconditionally link against hipblaslt causing clr to assert when these projects are used on a system that contains any other gpu or, alternatively, if assertions are disabled, causes clr to leave the code objects unloaded even for the supported devices in the system when a unsupported gpu is present, causing a crash when they are used.

### Operating System

Any linux

### CPU

Epyc 7552

### GPU

GFX900, GFX906, GFX908, GFX1030

### ROCm Version

ROCm 6.2.3

### Steps to Reproduce

Compile clr with assertions enabled

Have a system with a gpu not supported by hipblaslt
link any binary against hipblaslt.so (no need to use hipblaslt for anything)
observe assert when the binary is run.

**Comment by ppanchad-amd** [comment]
Hi @IMbackK. Internal ticket has been created to investigate your issue. Thanks!

**Comment by schung-amd** [comment]
Hi @IMbackK, thanks for keeping up with this. We're working on replacing this assertion and others with proper error handling. I'll reach out to the internal team to see if we have a timeline for this, as well as what the expected behavior will be in this usecase (i.e. heterogenous systems with one or more unsupported GPUs). Let me know if you have any additional questions and I'll pass those on as well.

**Comment by IMbackK** [comment]
Hi @schung-amd what is the sate of this? I see the code has changed significantly around this function, has this issue been adressed? what is the expected behavior on heterogeneous systems  at this time?

**Comment by schung-amd** [comment]
Hi @IMbackK, sorry for the delay on this. https://github.com/ROCm/clr/commit/019abdc3bd21028d10213631cc931c0487f4b272 should cover this issue, although I'm unsure when/if this made it into a release. With this commit an error will be returned instead of asserting when an invalid code object is loaded, but users are still responsible for handling this error properly. I'll see if I can find some info on what ROCm versions this is fixed in, if any.

**Comment by IMbackK** [comment]
> Hi [@IMbackK](https://github.com/IMbackK), sorry for the delay on this. [019abdc](https://github.com/ROCm/clr/commit/019abdc3bd21028d10213631cc931c0487f4b272) should cover this issue

Great

> With this commit an error will be returned instead of asserting when an invalid code object is loaded, but users are still responsible for handling this error properly.

Ok so when this is applied to a release this will break clients that indiscriminately link to hipblaslt but expect to function on devices which it dosent support, like pytorch and miopen. I do think this behavior is correct for the runtime, although handling the error in the clients is going to be a bit cumbersome as the error will be raised in ld before the clients have any chance to react, so they will have to dlopen instead.
I hope to find the time to try this out soon and file issues against the affected components if so.

**Comment by schung-amd** [comment]
Right, if the error isn't handled this won't be much different from the assertion. Thanks in advance for looking into possible failures if you get the time to do so. I'll close this for now, but feel free to comment here (as well as filing a report for the affected component) if you find any components which are still not handling this situation properly.
